### DISCUSSION TITLE
What need to be done to move Drasil beyond create software for other uses

### DISCUSSION THREAD

**Discussion by cd155** [JacquesCarette/Drasil#3235] [Ideas]
Recently, I was asked a question about how the Drasil framework can be used for full-stack development. By full-stack, I mean software has a front-end and a back-end. Usually, the front end can be a mobile application or a web portal. The back end is a server that usually handles relational databases. I actually wasn't sure how to answer this question.

By looking at the example we have, Projectile and NOPCM, we don't have the level of this design pattern. From what I can see, we have different modules which handle different tasks, for example, Input, Output, and Calculation. If we want to generate a back-end which handles a relational database, is that place we can add to? For example, add a module called "server code." Most of our examples don't have a front end; the front end is just a console, which we open the terminal and run the program. Currently, it doesn't have a server-end. All case studies run locally. So I was curious if we want to expand the Drasil framework to generate more genetic software, what will the roadmap look like? It seems a lot of work needs to be done, but I think it is an interesting topic to discuss because it stresses Drasil's capability to its limitation.

**Comment by peter-michalski** [comment]
There are many definitions of "front end" and "back end" but the existence of a back end does not necessitate the existence of a relational database.  Broadly, back end is what goes on 'behind the scenes", away from the user interface, usually server side, That is, if there even is a server, and there doesn't need to be. 

>We have different modules which handle different tasks, for example, Input, Output, and Calculation

I would say that in Drasil these modules are part of the back end.  

We use the terminal to run examples, but I would have a hard time calling it a front end. Drasil does not even prompt user input. Users just manually code their examples by adding information to all of the various example specific files within the example directory. There is no client side. Adding a front end to Drasil could involve developing some sort of GUI for users to fully specify their examples instead of what is manually done.

**Reply by cd155** [reply]
Thanks, Peter. I do agree with the definition you list for front-end and back-end. The person who asks this question has experience with full-stack development. 

I also think dividing the software into front-end, and back-end seem to follow the MVC (model, view, controller) design pattern. The front end serves as a viewer, and the back end serves as the controller and model. Currently, in Drasil, we can generate modularized software, and it seems promise we could add design pattern knowledge.
